codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@Scottmitch 

Agreed we want to keep this as simple as possible.  So how about this:

```java
interface Http2RemoteFlowController.FlowControlled {
    // Must not change the stream state. If an exception is thrown, error() will be called afterward.
    boolean write(int allowedBytes);

    // Called after this has been successfully written.
    // Only one call to this or error() will be made by the flow controller.
    void writeComplete();

    // Called to indicate that an error occurred before this object could be completely written.
    // Only one call to this or writeComplete will be made by the flow controller.
    void error(Throwable cause)
}```

Note that write can now throw an exception, which it would do rather than calling `error()` directly.

I think this should keep things simple and not require the nasty queue management.

WDYT?